Feat/coupon PR

ksah3756 opened the pull request:
## 주요 변경 사항

User 엔티티의 `user_coupon`이 user_id를 key로 하는 hashmap -> 다대다 관계이므로 같은 user_id가 여러개 있을 수 있으므로 key로는 사용 못하고, key를 사용할거면 user_coupon 별도의 unique id가 필요한데 이걸 키로 사용할거면 해시맵 사용이 의미가 없다고 생각..
```java
@ElementCollection(fetch = FetchType.LAZY)
    @CollectionTable(name = "user_coupon", joinColumns = @JoinColumn(name = "user_id"))
    private List<CouponEmbeddable> coupons = new ArrayList<>();
```
> erd 설계대로 coupon_id, coupon_name 갖도록 `CouponEmbeddable` 이라는 임베디드 타입 만들어서 사용해봄

비즈니스 로직이랑 나중에 락 처리 로직 추가할 경우 분리하기 위해 UserCouponService라는 클래스를 별도로 만들음 
https://velog.io/@myspy/Transaction-%EC%A0%84%ED%8C%8C%EA%B0%80-%EB%AD%A1%EB%8B%88%EA%B9%8C 이걸 보고 참고했습니다
> 근데 이게 맞는지 잘 모르겠음 어떤 락을 사용하느냐에 따라 다시 서비스 하나로 합칠 수도 있을 거 같음

기능 명세서대로 전체 쿠폰 조회, 유저 보유중인 쿠폰 조회, 쿠폰 유저 할당 3개의 api 구현하였고 테스트 코드 작성하였습니다

JiinHong committed:
feat: User 기능 추가

ksah3756 committed:
Add controller, service, repository, dto

ksah3756 committed:
쿠폰 api 완료, 동시성 처리 X

ksah3756 committed:
UserCouponService 주석 내용 변경

asdf56782 commented on src/main/java/com/example/goorm_ticket/domain/coupon/entity/CouponEmbeddable.java:
```diff
@@ -0,0 +1,35 @@
+@Embeddable
+@NoArgsConstructor
+@Getter
+public class CouponEmbeddable {
```
### @Embeddable 사용

Map<Long, String> 대신 @Embeddable 사용도 좋은 방법인 것 같습니다!!

*추가로 PR 코멘트에 'user_coupon이 user_id를 key로 하는' 이라고 작성해주셨는데, Map의 Key는 couponId를 의미합니다. 또한, Map<Long, String> 타입을 사용해도 coupon_id와 user_id로 복합키를 가지고 있긴 합니다!*

asdf56782 commented on src/main/java/com/example/goorm_ticket/api/coupon/service/UserCouponService.java:
```diff
@@ -0,0 +1,51 @@
+                Coupon coupon = couponRepository.findById(coupon_id).orElseThrow();
+                String coupon_name = coupon.getName();
+
+                userCoupons.add(CouponEmbeddable.builder()
```
### Builder 적용 위치

CouponEmbeddable에 대한 builder 작성은 CouponEmbeddable 도메인에 하는 것을 추천합니다!
Builder 패턴은 캡슐화, 관심사 분리, 가독성 등 여러 이유로 도메인에 작성하는 것을 권장한다고 알고 있습니다.

asdf56782 commented on src/test/java/com/example/goorm_ticket/api/coupon/service/CouponServiceTest.java:
```diff
@@ -0,0 +1,161 @@
+    private Long coupon2_id;
+
+    @BeforeEach
+    void before() {
```
### 객체 생성 메서드 생성

Coupon, CouponEmbeddable 등 객체를 생성하는 메서드를 따로 작성하면 중복되는 코드를 줄일 수 있을 것 같습니다!

ksah3756 replied:
of 메소드 작성완료했습니다

asdf56782 commented on src/main/java/com/example/goorm_ticket/api/coupon/service/UserCouponService.java:
```diff
@@ -0,0 +1,51 @@
+
+                userRepository.save(user);
+            }
+        } catch (IllegalArgumentException e) {
```
### 예외 상황에 맞는 응답 필요성

쿠폰 수량이 부족하면 IllegalArgumentException를 던져 해당 라인에서 잡고 있는데, 서버 내에서만 예외 메시지를 출력하는 것보다 **클라이언트에게도 '쿠폰 수량이 부족하다'는 응답 메시지를 보여주는 것**을 추천합니다!

✔️ **@ExceptionHandler**, 분기 생성 등 여러 방법을 이용하면 예외 상황에 맞는 응답을 별도로 작성할 수 있습니다.

asdf56782 replied:
System.out.print 말고 @Slf4j 사용!!

ksah3756 replied:
커스텀 예외 클래스 정의하였고 @ExceptionHandler 사용해서 예외 처리 완료하였습니다
sout 코드 삭제했습니다

ksah3756 committed:
refactor: 쿠폰 할당 메소드 이동

CouponController endpoint 수정
builder of 패턴 적
UserCouponService 삭제 후 쿠폰 할당 메소드 이동
쿠폰 관련 커스텀 예외 클래스 추가
글로벌 예외 처리 코드 추가
sout 코드 제거

ksah3756 committed:
fix: 설정파일 삭제

ksah3756 committed:
fix: 쿠폰 할당 api url 원래대로 변경

JiinHong committed:
수정사항 반영

wjd4204 commented on src/main/java/com/example/goorm_ticket/api/coupon/service/CouponService.java:
```diff
@@ -0,0 +1,88 @@
+
+    @Transactional(readOnly = true)
+    public List<CouponResponseDto> getUserCoupons(Long userId) {
+        User user = findUserById(userId);
```
userId를 찾지 못하였을 때의 exception 발행도 필요할 것 같습니다!

wjd4204 commented on src/main/java/com/example/goorm_ticket/api/coupon/service/CouponService.java:
```diff
@@ -0,0 +1,88 @@
+    }
+
+    public boolean decreaseCoupon(Long couponId) {
+        Coupon coupon = findCouponById(couponId);
```
해당 쿠폰아이디가 존재하지 않을 때의 exception을 추가하는 것이 좋아 보입니다!

wjd4204 commented on src/main/java/com/example/goorm_ticket/api/user/controller/UserController.java:
```diff
@@ -0,0 +1,43 @@
+import java.util.List;
+
+@RequiredArgsConstructor
+@RestController
```
마찬가지로 RequestMapping으로 공통 api domain을 만들어주세요

ksah3756 replied:
feat/User 브랜치에서 머지해서 푸시했습니다.

wjd4204 commented on src/main/java/com/example/goorm_ticket/domain/user/dto/UserOrderResponse.java:
```diff
@@ -0,0 +1,24 @@
+
+import java.time.LocalDateTime;
+
+public record UserOrderResponse(
```
Dto를 명명해 주세요.

ksah3756 replied:
feat/User 브랜치에서 머지해서 푸시했습니다

wjd4204 commented on src/main/java/com/example/goorm_ticket/domain/user/entity/User.java:
```diff
@@ -1,24 +1,22 @@
 
     @Id
-    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    @GeneratedValue(strategy = GenerationType.AUTO)
```
IDENTITY로 바꾸는 것이 좋을 것 같은데 AUTO로 하신 이유가 있나요?

ksah3756 replied:
feat/User 브랜치에서 머지해서 푸시했습니다.

wjd4204 commented on src/test/java/com/example/goorm_ticket/api/coupon/service/CouponServiceTest.java:
```diff
@@ -0,0 +1,156 @@
+    @Transactional
+    void 유저_쿠폰_조회() {
+        // given
+        List<CouponResponseDto> userCoupons = couponService.getUserCoupons(userId);
```
질문이 있습니다. getUserCoupons을 사용하는 것은 given보단 when이 맞지 않나요?

ksah3756 replied:
네 when이 맞는거같습니다 테스트 코드의 경우 전체적으로 수정해서 다시 올리겠습니다

ksah3756 committed:
merge feat/User
